### ISSUE TITLE
Read-only file system for config folder causes immediate crash. Unable to log due to same issue. (Caused by XDG BDS incompliance)

### ISSUE THREAD

**Issue by TheSast** [jesseduffield/lazygit#2856]
**Describe the bug**
Lazygit crashes after prompting for making a repo, before entering the TUI, if it cannot write to`$XDG_CONFIG_HOME/lazygit` and it is not possible to log the issue using `lazy --debug` if `$XDG_CONFIG_HOME/lazygit/development.log` is also not able to be written to.
Both of those are inevitable to happen if the lazygit config folder is in a read-only file system.

**To Reproduce**
This is not the only way it can happen but it is how I specifically encountered the issue.
Steps to reproduce the an instance of the behavior:

0. Install [home-manager](https://github.com/nix-community/home-manager), back up your lazygit config folder and run `home-manager init`.
1. Open `$HOME/.config/home-manager/home.nix` with a text editor.
2. Add `home.file.".config/lazygit/config.yml".text = "";` inside it's configuration.
3. Run `home-manager switch`
4. Run `lazygit` inside a git repository

**Expected behavior**
For program state files to respect `$XDG_STATE_HOME` as mentioned in #2794 with the expected default of `$HOME/.local/state`.
The XDG Base Directory Specification can be found [here](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

**Version info:**
This issue happens in the current lazygit nix package for nix stable 23.05 (which is lazygit 0.38.2) and on a build from the current latest commit of this repository [b92c294](https://github.com/jesseduffield/lazygit/commit/b92c29405930458dbaa6dbe242eb4576e3d462f2).
I have prepended the output of the `lazygit --version` command to the stderr output of lazygit of both builds.

**Stderr Output:**
`commit=, build date=, build source=nix, version=0.38.2, os=linux, arch=amd64, git version=2.40.1`
```
2023/07/31 21:16:03 An error occurred! Please create an issue at: https://github.com/jesseduffield/lazygit/issues

*fs.PathError open /home/u/.config/lazygit/state.yml: read-only file system
github.com/jesseduffield/lazygit/pkg/app/app.go:54 (0xab73dc)
github.com/jesseduffield/lazygit/pkg/app/entry_point.go:146 (0xab92da)
github.com/jesseduffield/lazygit/main.go:23 (0xabaebe)
runtime/internal/atomic/types.go:194 (0x439767)
runtime/asm_amd64.s:1598 (0x469b01)
```

`commit=, build date=, build source=unknown, version=unversioned, os=linux, arch=amd64, git version=2.40.1`
```
2023/07/31 21:31:33 An error occurred! Please create an issue at: https://github.com/jesseduffield/lazygit/issues

*fs.PathError open /home/u/.config/lazygit/state.yml: read-only file system
/home/u/go/pkg/mod/github.com/jesseduffield/lazygit@v0.39.4/pkg/app/app.go:55 (0xad7e9c)
	Run: newErr := errors.Wrap(err, 0)
/home/u/go/pkg/mod/github.com/jesseduffield/lazygit@v0.39.4/pkg/app/entry_point.go:151 (0xad9fa6)
	Start: Run(appConfig, common, appTypes.NewStartArgs(cliArgs.FilterPath, parsedGitArg, integrationTest))
/home/u/go/pkg/mod/github.com/jesseduffield/lazygit@v0.39.4/main.go:23 (0xadb85e)
	main: app.Start(ldFlagsBuildInfo, nil)
/nix/store/fp9k417pvah7d6ibbnz2x2llzframbk4-go-1.20.6/share/go/src/runtime/internal/atomic/types.go:194 (0x439767)
	(*Uint32).Load: return Load(&u.value)
/nix/store/fp9k417pvah7d6ibbnz2x2llzframbk4-go-1.20.6/share/go/src/runtime/asm_amd64.s:1598 (0x469b01)
	goexit: BYTE	$0x90	// NOP
```

**Comment by TheSast** [comment]
Similar to #818 with same changes necessary for resolution.

**Comment by stefanhaller** [comment]
I can't reproduce this with a read-only config folder (i.e. `chmod 555 "~/Library/Application Support/lazygit"` on Mac). Lazygit starts normally, and displays the "Thanks for using lazygit" message every time, because it can't record that it did so in `state.yml`. It seems that you really need a read-only file system to trigger the crash, but I'm too lazy to set one up to try that.

There's [code in app_config.go](https://github.com/jesseduffield/lazygit/blob/master/pkg/config/app_config.go#L276) that checks for permission errors when writing the state file and silently ignores those. My guess is that the `os.IsPermission` function doesn't check for `EROFS`, so we might consider including that check ourselves at that point.

**Comment by stefanhaller** [comment]
@TheSast This should be fixed by #2936, could you test that PR please?

**Comment by stefanhaller** [comment]
Thanks for testing!

> 1. It does fix the error only for when `state.yml` is absent from the config folder, it yelds same error as before if present, I assume it tries to move it to the new location,

No, it writes to the file again in the same location where it found it. Which means that for existing users it will continue to save state.yml in the same folder as config.yml, forever. I don't think that's a problem, since state.yml is only created by lazygit, so if it exists in the config folder, that folder must be writeable.

We did consider moving files to the new locations, but decided against it in #2936.

> 2. The log files are still written in the config folder making it impossible to check them for info.

Ah, I missed this one. Log files shouldn't be saved into the config folder. According to the XDG spec they should probably go to XDG_STATE_HOME. @horriblename, could you add that change to #2936? It looks like this will just be a one-line change to the `LogPath()` function at the end of `app_config.go`.

> But while testing I also encountered a similar issue for `config.yml`, lazygit will refuse to launch if it can't write `config.yml`.

Again, only on a read-only file system. I can't reproduce with a folder that doesn't have write permissions.

But anyway, since this is a slightly unrelated problem, I think it should be fixed in a different PR (if at all). It doesn't seem like a serious issue to me.

To sum it up, with the additional change of writing log files to XDG_STATE_HOME we're good in my opinion.

**Comment by horriblename** [comment]
> According to the XDG spec they should probably go to XDG_STATE_HOME. @horriblename, could you add that change to https://github.com/jesseduffield/lazygit/pull/2936?

log files belong in XDG_CACHE_HOME I think. Logs aren't part of the app state after all.

I can add it in #2936 but would probably be better to split it off, what do you think?

**Comment by horriblename** [comment]
seems reasonable to put it in XDG_STATE_HOME, will do that later in #2936
